collectors/payments_collector.go

```diff
@@ -0,0 +1,151 @@
+package collectors
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	totalPayments = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "lnd_total_payments",
+			Help: "Total number of payments initiated",
+		},
+	)
+	totalHTLCAttempts = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "lnd_total_htlc_attempts",
+			Help: "Total number of HTLC attempts across all payments",
+		},
+	)
+	paymentAttempts = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "lnd_payment_attempts_per_payment",
+			Help:    "Histogram tracking the number of attempts per payment",
+			Buckets: prometheus.LinearBuckets(1, 1, 10), // 1 to 10 attempts
+		},
+	)
+)
+
+// paymentsMonitor listens for payments and updates Prometheus metrics.
+type paymentsMonitor struct {
+	client routerrpc.RouterClient
+
+	ctx context.Context
+
+	errChan chan error
+
+	// quit is closed to signal that we need to shutdown.
+	quit chan struct{}
+
+	wg sync.WaitGroup
+}
+
+// newPaymentsMonitor creates a new payments monitor and ensures the context
+// includes macaroon authentication.
+func newPaymentsMonitor(lnd *lndclient.LndServices,
+	errChan chan error) (*paymentsMonitor, error) {
+
+	// Attach macaroon authentication for the router service.
+	ctx := context.Background()
+	ctx, err := lnd.WithMacaroonAuthForService(ctx,
+		lndclient.RouterServiceMac)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get macaroon-authenticated "+
+			"context: %w", err)
+	}
+
+	return &paymentsMonitor{
+		client:  routerrpc.NewRouterClient(lnd.ClientConn),
+		errChan: errChan,
+		quit:    make(chan struct{}),
+		ctx:     ctx,
+	}, nil
+}
+
+// start subscribes to `TrackPayments` and updates Prometheus metrics.
+func (p *paymentsMonitor) start() error {
+	// Use the stored authenticated context.
+	ctx, cancel := context.WithCancel(p.ctx)
+
+	paymentLogger.Info("Starting payments monitor...")
+
+	stream, err := p.client.TrackPayments(ctx,
```

Comment:
We should also verify that this is using: https://lightning.engineering/api-docs/api/lnd/router/track-payment-v2/

Reply:
That said, it might not be comparable as `TrackPaymentV2` is for a single payment whereas `TrackPayments` is for payment events in general

Reply:
I think we're good to go. Check this out: https://lightning.engineering/api-docs/api/lnd/router/track-payments/. Originally I was looking at using the _**lndclient**_ library like the rest of the code base, but it doesn't yet support _**TrackPayments**_ so decided to use raw _**routerrpc**_ client.